### PR TITLE
fix: "data collection for marketing" from PR #9687

### LINKED ISSUE #9862: fix: "data collection for marketing" toggles have bugs and redundant addTraitsToUser and events added in PR #9687
### What is this about?

https://github.com/MetaMask/metamask-mobile/pull/9687/ adds `addTraitsToUser` calls that should be grouped with the existing calls. Traits should be consolidated first then passed all at once as this makes the Segment SDK send an API call for identify.

In addition, passing traits to tracking events is redundant as values are already on the user profile.
This makes the `ANALYTICS_PREFERENCE_SELECTED` event itself redundant as it's then becomes similar to the already sent `ANALYTICS_PREFERENCE_SELECTED`.
The `location: 'onboarding_metametrics'` event prop is also a duplicate of `updated_after_onboarding: false` prop and we probably have to chose only one.

To be discussed as data team may have good reasons to need these data, but we should not just add them to the code as if nothing was already there...

### Scenario

_No response_

### Design

_No response_

### Technical Details

_No response_

### Threat Modeling Framework

_No response_

### Acceptance Criteria

_No response_

### Stakeholder review needed before the work gets merged

- [x] Engineering (needed in most cases)
- [ ] Design
- [ ] Product
- [ ] QA (automation tests are required to pass before merging PRs but not all changes are covered by automation tests - please review if QA is needed beyond automation tests)
- [ ] Security
- [ ] Legal
- [ ] Marketing
- [ ] Management (please specify)
- [x] Other (please specify)
  - data team @worldlyjohn  

### References

_No response_

### PR DESCRIPTION
**Pull request by NicolasMassart**
### Onboarding phase

- prevent metametrics API to be called multiple times for identical actions:
  - group traits addition
  - refactor events into a consolidated one
- remove useless trackEvent in places where it would never be triggered anyway (when metrics are disabled)
- add condition to prevent useless data processing when metrics are disabled
- move async work in `InteractionManager.runAfterInteractions` as hook doesn't do it by default on traits addition. Move the event too even if handled already to ensure consistency with traits addition.
- number of events reduced from 11 to 7 when marketing is opted-in.
- making checkbox text touchable to check the box
- add unit tests and update snapshot

### Settings screen

- extract a section component to make the section code testable as otherwise you have to mock everything and it conflicts with code under test
  but keep MetaMetrics toggle and marketing toggle together as they have been designed to work together closely.
- add traits when toggling marketing on/off
- rework MetaMetrics and Marketing switches interaction logic
- unit test on all the section and switches interactions

> [!NOTE]  
> This is a rework of #9687.
> I had to extract the settings section as a component for testing to be possible.
> I could have split it further into MetaMEtrics seciton and Marketing data section.
> but this would have required more work to make sure both components keep their linked behaviours when toggling switches.
> however, this is a possible improvement. I consider we should work on this on a full settings sections refactoring, see issue #9979

## **Related issues**

Fixes: #9862 

## **Manual testing steps**

- Same as for #9687
- No UI change.
- all changes on event tracking and identify are unit tested.

#### Onboaring phase

When not checking marketing checkbox
```
 INFO  TRACK event saved {"event": "Terms of Use Shown", "properties": {}, "type": "track"}
 INFO  IDENTIFY event saved {"traits": {"Batch account balance requests": "ON", "Enable OpenSea API": "ON", "NFT Autodetection": "OFF", "Theme": "light", "applicationVersion": "7.23.0", "currentBuildNumber": "1333", "deviceBrand": "Apple", "operatingSystemVersion": "17.5", "platform": "ios", "security_providers": "blockaid", "token_detection_enable": "ON"}, "type": "identify", "userId": "ff0d83a8-9c1e-4897-b1b1-1b5e5e23ed42"}
 INFO  TRACK event saved {"event": "Wallet Setup Started", "properties": {}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {"analytics_option_selected": "Metrics Opt In", "updated_after_onboarding": false}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {"analytics_option_selected": "Metrics Opt In", "updated_after_onboarding": false}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {}, "type": "track"}
 INFO  TRACK event saved {"event": "Welcome Message Viewed", "properties": {}, "type": "track"}
 INFO  TRACK event saved {"event": "Onboarding Started", "properties": {}, "type": "track"}
 INFO  Sent 1 events
 INFO  Sent 7 events
```
and Mixpanel view
image

When checking marketing checkbox
```
 INFO  TRACK event saved {"event": "Terms of Use Shown", "properties": {}, "type": "track"}
 INFO  IDENTIFY event saved {"traits": {"Batch account balance requests": "ON", "Enable OpenSea API": "ON", "NFT Autodetection": "OFF", "Theme": "light", "applicationVersion": "7.23.0", "currentBuildNumber": "1333", "deviceBrand": "Apple", "operatingSystemVersion": "17.5", "platform": "ios", "security_providers": "blockaid", "token_detection_enable": "ON"}, "type": "identify", "userId": "d1b09ff0-b550-4d7f-a894-1cdc13609473"}
 INFO  TRACK event saved {"event": "Wallet Setup Started", "properties": {}, "type": "track"}
 INFO  IDENTIFY event saved {"traits": {"has_marketing_consent": true, "is_metrics_opted_in": true}, "type": "identify", "userId": "d1b09ff0-b550-4d7f-a894-1cdc13609473"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {"has_marketing_consent": true, "is_metrics_opted_in": true, "location": "onboarding_metametrics"}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {"has_marketing_consent": true, "is_metrics_opted_in": true, "location": "onboarding_metametrics"}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {"analytics_option_selected": "Metrics Opt In", "updated_after_onboarding": false}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {"analytics_option_selected": "Metrics Opt In", "updated_after_onboarding": false}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {}, "type": "track"}
 INFO  TRACK event saved {"event": "Welcome Message Viewed", "properties": {}, "type": "track"}
 INFO  TRACK event saved {"event": "Onboarding Started", "properties": {}, "type": "track"}
 INFO  Sent 1 events
 INFO  Sent 11 events
```
and Mixpanel view:
image

#### Settings screen

Missing marketing setting events and user traits:

```
 INFO  IDENTIFY event saved {"traits": {"Batch account balance requests": "ON", "Enable OpenSea API": "ON", "NFT Autodetection": "OFF", "Theme": "light", "applicationVersion": "7.23.0", "currentBuildNumber": "1333", "deviceBrand": "Apple", "operatingSystemVersion": "17.5", "platform": "ios", "security_providers": "blockaid", "token_detection_enable": "ON"}, "type": "identify", "userId": "5e9545ed-84b3-4223-801e-d70fee12e10d"}
 INFO  TRACK event saved {"event": "Views Security & Privacy", "properties": {}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {"analytics_option_selected": "Metrics Opt in", "updated_after_onboarding": true}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {"analytics_option_selected": "Metrics Opt in", "updated_after_onboarding": true}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {}, "type": "track"}
 INFO  TRACK event saved {"event": "Views Security & Privacy", "properties": {}, "type": "track"}
 INFO  Sent 5 events
 INFO  Sent 1 events
```

#### Onboaring phase

When not checking marketing checkbox
```
 INFO  TRACK event saved {"event": "Terms of Use Shown", "properties": {}, "type": "track"}
 INFO  IDENTIFY event saved {"traits": {"Batch account balance requests": "ON", "Enable OpenSea API": "ON", "NFT Autodetection": "OFF", "Theme": "light", "applicationVersion": "7.23.0", "currentBuildNumber": "1333", "deviceBrand": "Apple", "is_metrics_opted_in": true, "operatingSystemVersion": "17.5", "platform": "ios", "security_providers": "blockaid", "token_detection_enable": "ON"}, "type": "identify", "userId": "19f4001c-68a6-4b62-9f1d-d8f6ff92718e"}
 INFO  TRACK event saved {"event": "Wallet Setup Started", "properties": {}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {"is_metrics_opted_in": true, "location": "onboarding_metametrics", "updated_after_onboarding": false}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {"is_metrics_opted_in": true, "location": "onboarding_metametrics", "updated_after_onboarding": false}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {}, "type": "track"}
 INFO  TRACK event saved {"event": "Welcome Message Viewed", "properties": {}, "type": "track"}
 INFO  TRACK event saved {"event": "Onboarding Started", "properties": {}, "type": "track"}
 INFO  Sent 1 events
 INFO  Sent 7 events
```
and Mixpanel view:
image

When checking marketing checkbox
```
 INFO  TRACK event saved {"event": "Terms of Use Shown", "properties": {}, "type": "track"}
 INFO  IDENTIFY event saved {"traits": {"Batch account balance requests": "ON", "Enable OpenSea API": "ON", "NFT Autodetection": "OFF", "Theme": "light", "applicationVersion": "7.23.0", "currentBuildNumber": "1333", "deviceBrand": "Apple", "has_marketing_consent": true, "is_metrics_opted_in": true, "operatingSystemVersion": "17.5", "platform": "ios", "security_providers": "blockaid", "token_detection_enable": "ON"}, "type": "identify", "userId": "fb576f56-3e5e-4df8-9a40-55d16ae5b16e"}
 INFO  TRACK event saved {"event": "Wallet Setup Started", "properties": {}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {"has_marketing_consent": true, "is_metrics_opted_in": true, "location": "onboarding_metametrics", "updated_after_onboarding": false}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {"has_marketing_consent": true, "is_metrics_opted_in": true, "location": "onboarding_metametrics", "updated_after_onboarding": false}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {}, "type": "track"}
 INFO  TRACK event saved {"event": "Welcome Message Viewed", "properties": {}, "type": "track"}
 INFO  TRACK event saved {"event": "Onboarding Started", "properties": {}, "type": "track"}
 INFO  Sent 1 events
 INFO  Sent 7 events
```
and mixpanel view:
image

#### Settings screen

```
 INFO  IDENTIFY event saved {"traits": {"Batch account balance requests": "ON", "Enable OpenSea API": "ON", "NFT Autodetection": "OFF", "Theme": "light", "applicationVersion": "7.23.0", "currentBuildNumber": "1333", "deviceBrand": "Apple", "is_metrics_opted_in": true, "operatingSystemVersion": "17.5", "platform": "ios", "security_providers": "blockaid", "token_detection_enable": "ON"}, "type": "identify", "userId": "5e9545ed-84b3-4223-801e-d70fee12e10d"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {"is_metrics_opted_in": true, "updated_after_onboarding": true}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {"is_metrics_opted_in": true, "updated_after_onboarding": true}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {}, "type": "track"}
 INFO  Sent 4 events
 INFO  IDENTIFY event saved {"traits": {"has_marketing_consent": true}, "type": "identify", "userId": "5e9545ed-84b3-4223-801e-d70fee12e10d"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {"has_marketing_consent": true, "location": "settings"}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {"has_marketing_consent": true, "location": "settings"}, "type": "track"}
 INFO  TRACK event saved {"event": "Analytics Preference Selected", "properties": {}, "type": "track"}
 INFO  Sent 4 events
```
and mixpanel view:
image

image

## **Pre-merge author checklist**

- [x] I’ve followed [MetaMask Contributor Docs](https://github.com/MetaMask/contributor-docs) and [MetaMask Mobile Coding Standards](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/CODING_GUIDELINES.md).
- [x] I've completed the PR template to the best of my ability
- [x] I’ve included tests if applicable
- [x] I’ve documented my code using [JSDoc](https://jsdoc.app/) format if applicable
- [x] I’ve applied the right labels on the PR (see [labeling guidelines](https://github.com/MetaMask/metamask-mobile/blob/main/.github/guidelines/LABELING_GUIDELINES.md)). Not required for external contributors.

## **Pre-merge reviewer checklist**

- [ ] I've manually tested the PR (e.g. pull and build branch, run the app, test code being changed).
- [ ] I confirm that this PR addresses all acceptance criteria described in the ticket it closes and includes the necessary testing evidence such as recordings and or screenshots.

### CODE REVIEW

**Comment by worldlyjohn** [comment]
@NicolasMassart i know this is still a WIP, but flagging for the "When not checking marketing checkbox" use case we should still send `has_marketing_consent` but with value of `false` (applies to both identify and track calls)

second question, in your examples why is `Analytics Preference Selected` fired 3 times?

**Review by jonybur — COMMENTED** [review]
Looks good!

**Review by jonybur — DISMISSED** [review]

**Review by jonybur — APPROVED** [review]

**Review by Cal-L — COMMENTED** [review]
Left some comments

**Review by Cal-L — APPROVED** [review]
LGTM

File: app/components/UI/OptinMetrics/index.js [hunk 1]
```diff
@@ -598,7 +595,15 @@ class OptinMetrics extends PureComponent {
                 : this.renderLegacyAction(action, i),
             )}
             {isPastPrivacyPolicyDate ? (
-              <View style={styles.checkbox}>
+              <TouchableOpacity
```
[the remark this answers is not in the document]
Comment (NicolasMassart) [reply app/components/UI/OptinMetrics/index.js:598]: This is for the text on the side to be touchable and trigger the checkbox as if you touched the box itself.

File: app/components/Views/Settings/SecuritySettings/Sections/MetaMetricsAndDataCollectionSection/MetaMetricsAndDataCollectionSection.tsx [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+import { setDataCollectionForMarketing } from '../../../../../../actions/security';
+import Routes from '../../../../../../constants/navigation/Routes';
+import { useMetrics } from '../../../../../hooks/useMetrics';
+import { useDispatch, useSelector } from 'react-redux';
+import { useNavigation } from '@react-navigation/native';
+
+const MetaMetricsAndDataCollectionSection: React.FC = () => {
+  const theme = useTheme();
+  const { colors } = theme;
+  const styles = createStyles(colors);
```
[the remark this answers is not in the document]
Comment (NicolasMassart) [reply app/components/Views/Settings/SecuritySettings/Sections/MetaMetricsAndDataCollectionSection/MetaMetricsAndDataCollectionSection.tsx:33]: Yes, sorry i just copied that from another settings section component

File: app/components/Views/Settings/SecuritySettings/Sections/MetaMetricsAndDataCollectionSection/MetaMetricsAndDataCollectionSection.tsx [hunk 3]
```diff
@@ -0,0 +1,186 @@
...
+
+  return (
+    <>
+      {renderMetaMetricsSection()}
+      {renderDataCollectionSection()}
+    </>
+  );
+};
+
+export default React.memo(MetaMetricsAndDataCollectionSection);
```
[the remark this answers is not in the document]
Comment (NicolasMassart) [reply app/components/Views/Settings/SecuritySettings/Sections/MetaMetricsAndDataCollectionSection/MetaMetricsAndDataCollectionSection.tsx:186]: Same, just copied from another settings component

**Comment by NicolasMassart** [comment]
@worldlyjohn 
> second question, in your examples why is `Analytics Preference Selected` fired 3 times?

See https://github.com/MetaMask/metamask-mobile/issues/9996

**Pull request merged**